src/Esprima/Token.cs

```diff
@@ -116,12 +118,13 @@ private sealed record TemplateHolder(string? Cooked, string RawTemplate, char No
     public readonly int LineNumber;
     public readonly int LineStart;
 
-    public object? Value => Type switch
+    public object? Value
     {
-        TokenType.Template => ((TemplateHolder) _value!).Cooked,
-        TokenType.RegularExpression => ((RegexHolder) _value!).Value,
-        _ => _value
-    };
+        // When this accessor gets inlined, it has a significant negative impact on max. nesting depth (on .NET 6),
+        // which is kind of weird...
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        get => _value is ValueHolder holder ? holder.Value : _value;
+    }
```

Comment:
@lahma Any idea on this? I had to add `MethodImplOptions.NoInlining` to pass the new max. depth test. I assume that the runtime started to inline the accessor automatically because it got shorter. I suspected that the local variable messed things up but changing it to something like `_value is ValueHolder ? ((ValueHolder) _value).Value : _value;` made no difference.

How come an inlined method uses more stack space than a non-inlined?

Reply:
Well at least this is now more complex for the runtime. Earlier was quick jump for regular case (no value holder) without type checks. Now this always incurs more expensive type check.

Reply:
Unfortunately, this doesn't answer the question: I tested the earlier version with forced inlining (`AggressiveInlining` applied to the getter of `Value`). The result is the same: inlining `Value` reduces max. nesting depth to around 430. It just hasn't turned out before because it seems that the runtime doesn't tend to inline the earlier version without explicitly telling it to do so, as opposed to the new one.

> Well at least this is now more complex for the runtime. Earlier was quick jump for regular case (no value holder) without type checks. Now this always incurs more expensive type check.

This, of course, is true but according to the above it may only affect execution time. I'll do some benchmarks later to ensure that we don't have a regression concerning that. We can revert to switch if that proves to be faster.

Reply:
While reading Pro .NET Memory Management yesterday, I noticed a small mention at some point suggesting that `AggressiveInlining` may actually disable a bunch of optimizations. It made me think of this. Any chance that forcing inlining disables layout optimizations of the struct, or similar?

Reply:
> I noticed a small mention at some point suggesting that AggressiveInlining may actually disable a bunch of optimizations

Basically, inlining should enable a lot of optimizations which aren't possible otherwise, not prevent them. Especially, on .NET 6: see [this excellent article](https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-6/#jit) by Stephen Toub on the topic.

I suppose the author of Pro .NET Memory Management is talking about this:

> There’s also the impact on the JIT compiler itself, as the JIT has limits on things like the size of a method before it’ll give up on optimizing further; inline too much code, and you can exceed said limits.

But this goes for the method *into* which we inline other methods, not for the methods we inline.

> Any chance that forcing inlining disables layout optimizations of the struct, or similar?

I doubt that inlining would affect the memory layout of structs. After some experimenting, I came to the conclusion that the problem lies around downcasting and accessing the `Value` property of the downcast holder object. After moving that part into a separate, non-inlined method, inlining started to work without extra stack consumption (see [my later comment](https://github.com/sebastienros/esprima-dotnet/pull/331#discussion_r956617418)). I haven't had the patience to unearth the exact reasons but that's for sure we're pushing the limits of the runtime's capabilities here.